### PR TITLE
Fixed the comparison about two integer as two unsigned integer

### PR DESCRIPTION
**Pull request by dinosaure**
In this PR, I believe than `bytes_received` and `bytes_written` can be an `unsigned int`. In this way, I don't find any problem about some current operations in Faraday. However, about the `flushes` queue, you compare these integers.

I put a comment to explain where is the problem (not a common problem but could be appear). This PR fixed this specific problem.

### CODE REVIEW

File: lib/faraday.ml [hunk 1]
```diff
@@ -422,7 +422,21 @@ let rec shift_buffers t written =
...
+
+       > bytes_written = max_int - 5
+       > Flushes.[ (max_int - 4, f); ... ]
+       > shift_buffers
+       > bytes_written += 10
+       > bytes_written >= threshold = false (but we expect [true])
+
+       This new kind of comparison takes care about semantic between unsigned
+       integers. *)
+    if compare (t.bytes_written - min_int) (threshold - min_int) >= 0
```
Comment (seliopou) [inline lib/faraday.ml]: Is there any reason why you're not just doing `t.bytes_written - min_int >= threshold - min_int`?

**Comment by seliopou** [comment]
If I'm reading this correctly, this is protecting against integer overflows? That's a potential concern on 32-bit architectures, but virtually impossible on a 64-bit architecture.

I'm not crazy about the additional operations, but everything should be in the cache or a register so probably doesn't matter much in practice? Let me think about it a bit more.

**Comment by dinosaure** [comment]
It's about which type is `bytes_received` and `bytes_written`. If you consider this `int` as an `int`, this PR has no reason to exist. But I believe you consider these values as `unsigned int`. As X. Leroy [said](http://caml.inria.fr/pub/ml-archives/caml-list/2004/07/f1c483068cc62075c916f7ad7d640ce0.fr.html), the operation like the addition does not matter if we consider an `int` as an `unsigned int`.

However, he said that comparisons, `div` and `mod` should not work as you want (if you expect, again, an `unsigned int`). This PR fixes only the comparison. This hack is come from [`mtime`](https://github.com/dbuenzli/mtime/blob/master/src/mtime.ml#L27).

**Comment by dinosaure** [comment]
And, in practice, this case could happen if we use `Faraday` as the server encoder (as a long life object). `bytes_received` and `bytes_written` could be close to `max_int` because you only add up these values.

```sh
λ rgrep bytes_written
lib/faraday.ml:  ; mutable bytes_written  : int
lib/faraday.ml:  ; bytes_written   = 0
lib/faraday.ml:  (t.write_pos - t.scheduled_pos) + (t.bytes_received - t.bytes_written)
lib/faraday.ml:    if compare (t.bytes_written - min_int) (threshold - min_int) >= 0
lib/faraday.ml:  t.bytes_written <- t.bytes_written + written;
```

```sh
λ rgrep bytes_received
lib/faraday.ml:  ; mutable bytes_received : int
lib/faraday.ml:  ; bytes_received  = 0
lib/faraday.ml:  t.bytes_received <- t.bytes_received + len;
lib/faraday.ml:  else Flushes.enqueue (t.bytes_received, f) t.flushed
lib/faraday.ml:  (t.write_pos - t.scheduled_pos) + (t.bytes_received - t.bytes_written)
lib/faraday.ml:       [t.bytes_received] and [threshold]). However, if [t.bytes_written] will
```

**Comment by seliopou** [comment]
So, I thought about it a little more. This can't be said to fix the overflow problem without making additional assumptions about the _relative rate_ at which the two quantities being compared are changing. Put another way, this change allows the difference between the two quantities to be twice as large without running into any problems due to overflows.

That's certainly an improvement for 32 bit architectures but on 64 bit architectures I don't think this will come up much. But, I don't think it would hurt to have this as a backup policy in either case.

**Pull request merged**